bf16-bitfury16.c

```diff
@@ -0,0 +1,1190 @@
+#include <arpa/inet.h>
+
+#include "miner.h"
+
+#include "bf16-bitfury16.h"
+#include "bf16-communication.h"
+#include "bf16-ctrldevice.h"
+
+#include "math.h"
+#include "sha2.h"
+#include "driver-bitfury16.h"
+
+//#define FLIP_BITS
+
+bf_cmd_description_t cmd_description[CHIP_CMD_NUM] = {
+	{
+		.cmd_code        = CHIP_CMD_TASK_STATUS,
+		.cmd_description = "CHIP_CMD_TASK_STATUS"
+	},
+	{
+		.cmd_code        = CHIP_CMD_TASK_WRITE,
+		.cmd_description = "CHIP_CMD_TASK_WRITE"
+	},
+	{
+		.cmd_code        = CHIP_CMD_TASK_SWITCH,
+		.cmd_description = "CHIP_CMD_TASK_SWITCH"
+	},
+	{
+		.cmd_code        = CHIP_CMD_READ_NONCE,
+		.cmd_description = "CHIP_CMD_READ_NONCE"
+	},
+	{
+		.cmd_code        = CHIP_CMD_SET_CLOCK,
+		.cmd_description = "CHIP_CMD_SET_CLOCK"
+	},
+	{
+		.cmd_code        = CHIP_CMD_TOGGLE,
+		.cmd_description = "CHIP_CMD_TOGGLE"
+	},
+	{
+		.cmd_code        = CHIP_CMD_SET_MASK,
+		.cmd_description = "CHIP_CMD_SET_MASK"
+	},
+	{
+		.cmd_code        = CHIP_CMD_CREATE_CHANNEL,
+		.cmd_description = "CHIP_CMD_CREATE_CHANNEL"
+	}
+};
+
+char* get_cmd_description(bf_cmd_code_t cmd_code) {
+	uint8_t i;
+	for (i = 0; i < CHIP_CMD_NUM; i++)
+		if (cmd_description[i].cmd_code == cmd_code)
+			return cmd_description[i].cmd_description;
+
+	return NULL;
+}
+
+static void shift_bits(uint8_t* data, uint8_t size, uint8_t nbits) {
+	uint8_t i;
+	uint8_t bytes = nbits / 8;
+	uint8_t bits  = nbits % 8;
+
+	for (i = 0; i < size; i++) {
+		data[i] = (data[i + bytes] << bits);
+		uint8_t minor = (data[i + bytes + 1] >> (8 - bits));
```

Comment:
This one is fixed
I did not check other code - perhaps some more revision is needed regarding styling...